### PR TITLE
Traitors see codewords highlighted

### PR DESCRIPTION
**Pull request by Kyep**
WIP PR that highlights codewords in radio messages and local chat heard by fellow syndicate agents.

Essentially, it makes codewords obvious, so if another traitor tries to use them to tip you off, you'll notice.
Obviously, doesn't make the words appear any differently for normal crew. Only traitors see the special formatting.

Demo: http://i.imgur.com/2M00bQr.png
Thread: https://nanotrasen.se/phpBB3/viewtopic.php?f=12&t=9402

### CODE REVIEW

**Comment by tigercat2000** [comment]
Why is this necessary? If it's to help people that care about codewords, they can set the word highlighting in Goonchat. If it's to encourage people that don't care about codewords to care about codewords, they're unlikely to notice bolded words in say, or just ignore it.

**Comment by BiberDark** [comment]
Nobody knows how to set a word highlighting in goonchat. For example me. And second you have to actual do something on yourselve, with this PR it is automatic and nobody will overlook it again. Good for Newbies too. If i read the forum correct, they are not only bold but in a separate color. Very flashy.

**Comment by MarsM0nd** [comment]
About the Goonchat highlighting beeing able to take over this feature: As far as i know, you can only highlight 5 strings, and if you have 6 codewords, 3 as start, 3 as response, you can't mark them all, and you can't mark what diffrent stuff during it, for example your characters name, to easily see if someone did address you.

File: code/controllers/master_controller.dm [hunk 1]
```diff
@@ -38,8 +40,15 @@ var/global/pipe_processing_killed = 0
 		job_master.LoadJobs("config/jobs.txt")
 		log_startup_progress("Job setup complete in [stop_watch(watch)]s.")
 
-	if(!syndicate_code_phrase)		syndicate_code_phrase	= generate_code_phrase()
-	if(!syndicate_code_response)	syndicate_code_response	= generate_code_phrase()
+	if(!syndicate_code_phrase)
+		var/list/new_phrases = generate_code_phrase()
+		syndicate_code_phrase = n_jointext(new_phrases, ", ") + "."
```
Comment (Krausus) [inline code/controllers/master_controller.dm]: `n_jointext` is just a `call`able version of the built-in `jointext`, which you should use directly in normal code.

File: code/controllers/master_controller.dm [hunk 2]
```diff
@@ -38,8 +40,15 @@ var/global/pipe_processing_killed = 0
...
 
-	if(!syndicate_code_phrase)		syndicate_code_phrase	= generate_code_phrase()
-	if(!syndicate_code_response)	syndicate_code_response	= generate_code_phrase()
+	if(!syndicate_code_phrase)
+		var/list/new_phrases = generate_code_phrase()
+		syndicate_code_phrase = n_jointext(new_phrases, ", ") + "."
+		syndicate_codewords += new_phrases
+	if(!syndicate_code_response)
+		var/list/new_responses = generate_code_phrase()
+		syndicate_code_response = n_jointext(new_responses, ", ") + "."
```
Comment (Krausus) [inline code/controllers/master_controller.dm]: `jointext` again.

**Comment by KasparoVy** [comment]
I mentioned what BiberDark and tigercat2000 said in the issue report for this as well, however I believe MarsM0nd's point is quite valid.

I don't see why this _shouldn't_ be a thing.

**Comment by Fox-McCloud** [comment]
I was in support of this until @KasparoVy pointed out what he did; not I'm definitively against it.

When a feature exists and isn't being used, it calls for education and spreading of that feature and how to use it, not coding and putting in a redundant feature on top of it.

With how the goonchat strings are handled, you can put all of the codewords in a single string, and, to my knowledge, it'll pick them out, individually, as opposed to if they're all together.

**Comment by Krausus** [comment]
It looks like Prisoner Processing didn't get highlighted in the demo screenshot, presumably because the area name includes the `\improper` tag. We don't have a proc for stripping those, but Bay does it like this:
```
#define strip_improper(input_text) replacetext(replacetext(input_text, "\proper", ""), "\improper", "")
```
I'm *pretty sure* BYOND omits the space that comes after an `\improper`, but you might want to double-check just in case you also need to trim to get rid of it.

**Comment by Lady-Luck** [comment]
This really needs to happen. I've never had a codeword work even once as a traitor, and I drop them every time.

Expecting everyone to know how the goonchat feature works on top of them remembering to even use the codeword system is extremely unrealistic.

**Comment by Kyep** [comment]
So, options:
A) Do nothing. Leave codewords as essentially useless.
B) Merge as-is. Ignores goonchat, but makes codewords work.
C) Replace with system that PROMPTS (but does not automatically) add traitor codewords to goonchat highlight. IMHO, this would be useless as most people would not do it. I wasn't even aware goonchat highlight existed until someone pointed it out on the forum recently. 
D) Replace with system that AUTO-ADDS traitor codewords to your goonchat highlight if you're a traitor. People seem to agree this is the best solution. But... I don't know how to do it - especially since the goonchat systems seems to be almost entirely client-side JS. I'm not even sure the server can tell you what to highlight or not.

**Comment by BiberDark** [comment]
Option B !

**Comment by MarsM0nd** [comment]
I vote for option C. People then can decide to have the highlighting or not, they might want to not have the highlighting as a challenge for themselfs, for example, and the prompt would also explain an existing feature, that might get used more thanks to it. I didn't know the part that you can 5 strings, with probably an unlimited ammount of words to search for in it, not only 5 strings that get checked for, only thanks to the traitor highlighting i know it now.

**Comment by Crazylemon64** [comment]
There's some `ehjax` thing with goonchat lemme investigate it - it might be what we need

**Comment by AndriiYukhymchak** [comment]
I use highlights ALWAYS, but I never attach codewords to there... muh slots

and yes, I know it treats spaces as separate... and I hate it sooo much

**Comment by DesolateG** [comment]
B

**Comment by Kyep** [comment]
If anyone else wants to take this on, ehjaxCallback in browserOutput.js might be used to pass instructions to the client as to which words they should highlight. 
Crazylemon did some experimentation in this regard: https://github.com/Crazylemon64/Paradise/commit/9963d0abb444d8a8f49632ff5c559653cb1d0636

**Comment by Lady-Luck** [comment]
Could this get resurrected due to #6102?

**Pull request closed without merging**